kong/router/fields.lua

```diff
@@ -0,0 +1,317 @@
+local buffer = require("string.buffer")
+local context = require("resty.router.context")
+
+
+local type = type
+local pairs = pairs
+local ipairs = ipairs
+local assert = assert
+
+
+-- cache key string
+local str_buf = buffer.new(64)
+
+
+local is_http = ngx.config.subsystem == "http"
+
+
+local MATCH_CTX_FUNCS
+local CACHE_KEY_FUNCS
+local get_cache_key
+
+
+if is_http then
+
+
+local tb_sort = table.sort
+local tb_concat = table.concat
+local replace_dashes_lower = require("kong.tools.string").replace_dashes_lower
+
+
+CACHE_KEY_FUNCS = {
+    ["http.method"] =
+    function(v, params, buf)
+      buf:put(params.method or ""):put("|")
+    end,
+
+    ["http.path"] =
+    function(v, params, buf)
+      buf:put(params.uri or ""):put("|")
+    end,
+
+    ["http.host"] =
+    function(v, params, buf)
+      buf:put(params.host or ""):put("|")
+    end,
+
+    ["tls.sni"] =
+    function(v, params, buf)
+      buf:put(params.sni or ""):put("|")
+    end,
+
+    ["http.headers."] =
+    function(v, params, buf)
+      local headers = params.headers
+      if not headers then
+        return
+      end
+
+      for _, name in ipairs(v) do
+        local name = replace_dashes_lower(name)
+        local value = headers[name]
+
+        if type(value) == "table" then
+          tb_sort(value)
+          value = tb_concat(value, ",")
+        end
+
+        buf:putf("%s=%s|", name, value)
+      end -- for ipairs(v)
+    end,
+
+    ["http.queries."] =
+    function(v, params, buf)
+      local queries = params.queries
+      if not queries then
+        return
+      end
+
+      for _, name in ipairs(v) do
+        local value = queries[name]
+
+        if type(value) == "table" then
+          tb_sort(value)
+          value = tb_concat(value, ",")
+        end
+
+        buf:putf("%s=%s|", name, value)
+      end -- for ipairs(v)
+    end,
+}
+
+
+MATCH_CTX_FUNCS = {
+    ["http.method"] =
+    function(v, c, params)
+      return c:add_value("http.method", params.method)
+    end,
```

Comment:
This seems to be a bit repetitive with the function above. Can we `add_value` and append cache key at the same time? This will reduce the amount of duplicate code and be slightly faster.

Reply:
The calculation of cache key happens earlier (`exec()`) than context matching (`select()`), 
I think that we can not do that in current workflow.

Reply:
We can operate like this:

```
provide_field_value(function(field, value)
  buf:put(value):put("|")
end)

provide_field_value(function(field, value)
  c:add_value(field, value)
end)
```

Or something similar, this should make the code reusable without sacrificing much performance.

Reply:
done.